drivers/serial/uart_mcux_iuart.c

```diff
@@ -0,0 +1,332 @@
+{
+	const struct mcux_iuart_config *config = DEV_CFG(dev);
+	int err = 0;
+
+	if (UART_GetStatusFlag(config->base, kUART_RxOverrunFlag)) {
+		err |= UART_ERROR_OVERRUN;
+		UART_ClearStatusFlag(config->base, kUART_RxOverrunFlag);
+	}
+
+	if (UART_GetStatusFlag(config->base, kUART_ParityErrorFlag)) {
+		err |= UART_ERROR_PARITY;
+		UART_ClearStatusFlag(config->base, kUART_ParityErrorFlag);
+	}
+
+	if (UART_GetStatusFlag(config->base, kUART_FrameErrorFlag)) {
+		err |= UART_ERROR_FRAMING;
+		UART_ClearStatusFlag(config->base, kUART_FrameErrorFlag);
+	}
+
+	return err;
+}
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+static int mcux_iuart_fifo_fill(struct device *dev, const uint8_t *tx_data,
+			       int len)
+{
+	const struct mcux_iuart_config *config = DEV_CFG(dev);
+	uint8_t num_tx = 0U;
+
+	while ((len - num_tx > 0) &&
+	       (UART_GetStatusFlag(config->base, kUART_TxEmptyFlag))) {
+
+		UART_WriteByte(config->base, tx_data[num_tx++]);
+	}
+
+	return num_tx;
+}
+
+static int mcux_iuart_fifo_read(struct device *dev, uint8_t *rx_data,
+			       const int len)
+{
+	const struct mcux_iuart_config *config = DEV_CFG(dev);
+	uint8_t num_rx = 0U;
+
+	while ((len - num_rx > 0) &&
+	       (UART_GetStatusFlag(config->base, kUART_RxDataReadyFlag))) {
+
+		rx_data[num_rx++] = UART_ReadByte(config->base);
+	}
+
+	return num_rx;
+}
+
+static void mcux_iuart_irq_tx_enable(struct device *dev)
+{
+	const struct mcux_iuart_config *config = DEV_CFG(dev);
+
+	UART_EnableInterrupts(config->base, kUART_TxEmptyEnable);
+}
+
+static void mcux_iuart_irq_tx_disable(struct device *dev)
+{
+	const struct mcux_iuart_config *config = DEV_CFG(dev);
+
+	UART_DisableInterrupts(config->base, kUART_TxEmptyEnable);
+}
+
+static int mcux_iuart_irq_tx_complete(struct device *dev)
+{
+	const struct mcux_iuart_config *config = DEV_CFG(dev);
+
+	return (UART_GetStatusFlag(config->base, kUART_TxEmptyFlag)) != 0U;
+}
+
+static int mcux_iuart_irq_tx_ready(struct device *dev)
+{
+	const struct mcux_iuart_config *config = DEV_CFG(dev);
+	uint32_t mask = kUART_TxEmptyEnable;
+
+	return (UART_GetEnabledInterrupts(config->base) & mask)
+		&& mcux_iuart_irq_tx_complete(dev);
+}
+
+static void mcux_iuart_irq_rx_enable(struct device *dev)
+{
+	const struct mcux_iuart_config *config = DEV_CFG(dev);
+	uint32_t mask = kUART_RxDataReadyEnable;
+
+	UART_EnableInterrupts(config->base, mask);
+}
+
+static void mcux_iuart_irq_rx_disable(struct device *dev)
+{
+	const struct mcux_iuart_config *config = DEV_CFG(dev);
+	uint32_t mask = kUART_RxDataReadyEnable;
+
+	UART_DisableInterrupts(config->base, mask);
+}
+
+static int mcux_iuart_irq_rx_full(struct device *dev)
+{
+	const struct mcux_iuart_config *config = DEV_CFG(dev);
+
+	return (UART_GetStatusFlag(config->base, kUART_RxDataReadyFlag)) != 0U;
+}
+
+static int mcux_iuart_irq_rx_ready(struct device *dev)
+{
+	const struct mcux_iuart_config *config = DEV_CFG(dev);
+	uint32_t mask = kUART_RxDataReadyFlag;
```

Comment:
The kUART_RxDataReadyFlag should be kUART_RxDataReadyEnable.
I have tested this with the uart shell.